File: leetcode_java/2006.CountNumberOfPairsWithAbsDifferenceK.java
package leetcode_java;

import java.util.HashMap;

class Test {
    public static void main(String[] args) {
        System.out.println(new Solution().countKDifference(new int[] {1,2,2,1}, 1));
    }
}

class Solution {
    public int countKDifference(int[] nums, int k) {
        int count = 0;
        HashMap<Integer, Integer> freq = new HashMap<>();

        // count occurance of each number
        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }

        for (int num : freq.keySet()) {
            if (freq.keySet().contains(num + k)) {
                count += freq.get(num) * freq.get(num + k);
            }
        }
        return count;
    }
}

File: 2006.Count_number_of_pairs_with_abs_difference_k.py
from typing import List

# EASY
class Solution:
    def countKDifference(self, nums: List[int], k: int) -> int:
        ## APPROACH: Brute Force
        ## TIME: O(n * n)
        ## Space: O(1)
        '''
        counter, l = 0, len(nums)

        for i in range(l):
            for j in range(i + 1, l):
                if abs(nums[i] - nums[j]) == k: counter += 1

        return counter
        '''
        ## APPROACH: frequency/occurance using hashmap
        ## TIME: O(n)
        ## Space: O(1)

        count = 0
        freq = {}

        for n in nums:
            # count the number of times a number appears
            freq[n] = freq.get(n,0) + 1
        # print(freq)
        for n in freq:
            if n + k in freq:
                count += freq[n] * freq[n + k]
        return count
        


obj = Solution()

nums, k = [1,2,2,1], 1  # o/p = 4
# nums, k = [1,3], 3      # o/p = 0
# nums, k = [2,2,3,1], 1  # o/p = 4

print(obj.countKDifference(nums, k))